lib/spack/spack/solver/asp.py

```diff
@@ -1263,36 +1316,53 @@ def provider_requirements(self):
         assert self.possible_virtuals is not None, msg
         for virtual_str in sorted(self.possible_virtuals):
             requirements = packages_yaml.get(virtual_str, {}).get("require", [])
-            rules = self._rules_from_requirements(virtual_str, requirements)
-            self.emit_facts_from_requirement_rules(rules, virtual=True)
+            rules = self._rules_from_requirements(
+                virtual_str, requirements, provenance=Provenance.PACKAGES_YAML_VIRTUAL
```

Comment:
> Is it for more than identifying rules as having come from virtuals in emit_facts_from_requirement_rules? 

No, it's for that. Here I package rules upfront in a data structure and process them afterwards. Why do you think the use is overloaded? What would you prefer instead? `VersionProvenance` vs. `RequirementProvenance`?

Reply:
> Why do you think the use is overloaded?

Provenance is intended as a component of multi-dimensional weighting (i.e. something that ends up as part of the facts in the final program). If you are just using it to change behavior outside of the program (when in fact the weights don't need to differ) then you are getting an unintended side effect as part of using this to influence other behavior. 

> What would you prefer instead? VersionProvenance vs. RequirementProvenance?

I would prefer not using the term "provenance", because idiomatically that is used in this code for weighting. I think it should be another property of a `RequirementRule`.

Reply:
I can code another enum, it needs to distinguish between "package" labels (e.g. `hdf5`), the label "all" and "virtual" labels (e.g. `mpi`).